chi/hardware.py

```diff
@@ -0,0 +1,155 @@
+from dataclasses import dataclass
+from datetime import datetime, timezone
+from typing import List, Optional, Tuple
+
+from .clients import blazar
+from .context import get, RESOURCE_API_URL
+
+import requests
+import logging
+
+LOG = logging.getLogger(__name__)
+
+@dataclass
+class Node:
+    """
+    Represents the Chameleon hardware that goes into a single node.
+    A dataclass for node information directly from the hardware browser.
+    """
+    site: str
+    name: str
+    type: str
+    architecture: dict
+    bios: dict
+    cpu: dict
+    gpu: dict
+    main_memory: dict
+    network_adapters: List[dict]
+    placement: dict
+    storage_devices: List[dict]
+    uid: str
+    version: str
+
+    def next_free_timeslot(self) -> datetime:
+        """
+        (Not implemented yet) Finds the next available timeslot for the hardware.
+
+        Returns:
+            A tuple containing the start and end datetime of the next available timeslot.
+            If no timeslot is available, returns the end datetime of the last lease.
+
+        """
+        raise NotImplementedError()
+
+        blazarclient = blazar()
+
+        # Get all leases
+        leases = blazarclient.lease.list()
```

Comment:
This should look at allocations, rather than leases. Leases will only show the current user's leases.

Here is the `get_allocations` function in the client:
https://github.com/ChameleonCloud/python-blazarclient/blob/chameleoncloud/xena/blazarclient/v1/hosts.py#L56

Here is the API spec for blazar, with the structure of the data `get_allocations` should return 
https://docs.openstack.org/api-ref/reservation/v1/index.html#get-allocation